### PR TITLE
PP-10077 Perform 3DS Flex device data collection for Google Pay

### PR DESCRIPTION
**Pull request by SandorArpa**
## WHAT
If appropriate, perform Worldpay 3DS Flex device data collection for Google Pay payments.

with @SandorArpa and @iqbalgds

### CODE REVIEW

File: app/assets/javascripts/browsered/web-payments/google-pay.js [hunk 1]
```diff
@@ -39,6 +44,95 @@ const processPayment = paymentData => {
...
+  }
+}
+
+const processPayment = paymentData => {
+  toggleSubmitButtons()
+  showSpinnerAndHideMainContent()
+
+  if (typeof Charge.worldpay_3ds_flex_ddc_jwt === 'string' && Charge.worldpay_3ds_flex_ddc_jwt !== '') {
+    const cachedDdcResult = window.Charge.googlePayWorldpay3dsFlexDeviceDataCollectionResult
+    if (!cachedDdcResult || !(typeof googlePayDDCEnabled === 'string')) {
```
Comment (kbottla) [inline app/assets/javascripts/browsered/web-payments/google-pay.js:126]: Looks like `googlePayDDCEnabled` is not defined anywhere and is not needed?

File: app/assets/javascripts/browsered/web-payments/google-pay.js [hunk 2]
```diff
@@ -39,6 +44,95 @@ const processPayment = paymentData => {
...
+  const deviceDataCollectionTimeout = window.setTimeout(() => {
+    window.removeEventListener('message', deviceDataCollectionFinishedEventListener)
+    submitGooglePayAuthRequest(paymentData, `DDC timeout after ${DDC_TIMEOUT_IN_MILLISECONDS} milliseconds`)
+  }, DDC_TIMEOUT_IN_MILLISECONDS)
+
+  var iframe = document.getElementById('googlePayWorldpay3dsFlexDdcIframe')
+
+  try {
+    // eslint-disable-next-line no-unused-expressions
+    iframe.contentWindow.location.href
```
Comment (kbottla) [inline app/assets/javascripts/browsered/web-payments/google-pay.js:85]: What is this expected to do?

File: app/assets/javascripts/browsered/web-payments/google-pay.js [hunk 3]
```diff
@@ -39,6 +44,95 @@ const processPayment = paymentData => {
...
+const checkOriginHostName = origin => {
+  return Charge.worldpay_3ds_flex_ddc_url.lastIndexOf(origin, 0) === 0
+}
+
+const performDeviceDataCollectionForGooglePay = (paymentData) => {
+  const deviceDataCollectionFinishedEventListener = event => {
+    if (checkOriginHostName(event.origin) && event.source === iframeWindow) {
+      const { MessageType, SessionId, Status } = JSON.parse(event.data)
+      if (MessageType === 'profile.completed') {
+        var ddcStatus = null
```
Comment (kbottla) [inline app/assets/javascripts/browsered/web-payments/google-pay.js:56]: as this is always set, no need to initalise 

```suggestion
        var ddcStatus
```

File: app/assets/javascripts/browsered/web-payments/google-pay.js (hunk 2 above)
Comment (SandorArpa) [reply app/assets/javascripts/browsered/web-payments/google-pay.js:85]: Same-origin locks access to href. When trying to access href and it throws an exception, then we know that we have to re-create the iframe as it has been changed by Google.

File: app/assets/javascripts/browsered/web-payments/google-pay.js (hunk 1 above)
Comment (SandorArpa) [reply app/assets/javascripts/browsered/web-payments/google-pay.js:126]: 👍

File: app/assets/javascripts/browsered/web-payments/google-pay.js (hunk 3 above)
Comment (SandorArpa) [reply app/assets/javascripts/browsered/web-payments/google-pay.js]: 👍

File: app/assets/javascripts/browsered/web-payments/google-pay.js (hunk 2 above)
Comment (kbottla) [reply app/assets/javascripts/browsered/web-payments/google-pay.js:85]: Do you mean changed by DDC? Maybe add a comment to explain it?

File: app/assets/javascripts/browsered/web-payments/google-pay.js [hunk 4]
```diff
@@ -39,6 +44,95 @@ const processPayment = paymentData => {
...
+  return Charge.worldpay_3ds_flex_ddc_url.lastIndexOf(origin, 0) === 0
+}
+
+const performDeviceDataCollectionForGooglePay = (paymentData) => {
+  const deviceDataCollectionFinishedEventListener = event => {
+    if (checkOriginHostName(event.origin) && event.source === iframeWindow) {
+      const { MessageType, SessionId, Status } = JSON.parse(event.data)
+      if (MessageType === 'profile.completed') {
+        var ddcStatus
+        var ddcResult
```
Comment (kbottla) [inline app/assets/javascripts/browsered/web-payments/google-pay.js:57]: `ddcResult` won't be set all the time and can be initialised.

Comment (SandorArpa) [reply app/assets/javascripts/browsered/web-payments/google-pay.js]: 👍

File: app/assets/javascripts/browsered/web-payments/google-pay.js (hunk 2 above)
Comment (SandorArpa) [reply app/assets/javascripts/browsered/web-payments/google-pay.js:87]: 👍

**Review by kbottla — APPROVED** [review]

**Pull request merged**